utilities/transactions/write_prepared_transaction_test.cc

```diff
@@ -773,6 +773,154 @@ TEST_P(WritePreparedTransactionTest, MaybeUpdateOldCommitMap) {
+    }
+    uint64_t num_imm_mems;
+    ASSERT_TRUE(db->GetIntProperty(DB::Properties::kNumImmutableMemTable,
+                                       &num_imm_mems));
+    if (attempt == kAttemptHistoryMemtable) {
+      ASSERT_EQ(0, num_imm_mems);
+    } else {
+      assert(attempt == kAttemptImmMemTable);
+      ASSERT_EQ(1, num_imm_mems);
+    }
+    
+    // Put something in active memtable
+    ASSERT_OK(db->Put(write_options, Slice("foo3"), Slice("bar")));
+
+    // Create txn3 after flushing, but this transaction also needs to
+    // check all memtables because of they contains uncommitted data.
+    Transaction* txn3 = db->BeginTransaction(write_options);
+    ASSERT_TRUE(txn3 != nullptr);
+    ASSERT_OK(txn3->SetName("txn3"));
+    // Write to the transaction just to ankor a snapshot.
+    txn3->SetSnapshotOnNextOperation();
+    ASSERT_OK(txn3->Put("foofoo3", "bar"));
+
+    // Commit the pending write
+    ASSERT_OK(txn_x->Commit());    
+
+    // Commit txn, txn2 and tx3. txn and tx3 will conflict but txn2 will
+    // pass. In all cases, both memtables are queried.
+    SetPerfLevel(PerfLevel::kEnableCount);
+    get_perf_context()->Reset();
+    ASSERT_TRUE(txn3->GetForUpdate(read_options, "foo", &value).IsBusy());
+    // We should have checked two memtables
+    ASSERT_EQ(2, get_perf_context()->get_from_memtable_count);
+    
+    get_perf_context()->Reset();
+    ASSERT_TRUE(txn->GetForUpdate(read_options, "foo", &value).IsBusy());
+    // We should have checked two memtables
+    ASSERT_EQ(2, get_perf_context()->get_from_memtable_count);
+
+    get_perf_context()->Reset();
+    ASSERT_OK(txn2->GetForUpdate(read_options, "foo2", &value));
+    ASSERT_EQ(value, "bar"); 
+    // We should have checked two memtables, and since there is not
+    // conflict, another Get() will be made and fetch the data from
+    // DB. If it is in immutable memtable, two extra memtable reads
+    // will be issued. If it is not (in history), only one will
+    // be made.
+    if (attempt == kAttemptHistoryMemtable) {
+      ASSERT_EQ(3, get_perf_context()->get_from_memtable_count);
+    } else {
+      assert(attempt == kAttemptImmMemTable);
+      ASSERT_EQ(4, get_perf_context()->get_from_memtable_count);
+    }
+
+    Transaction* txn4 = db->BeginTransaction(write_options);
+    ASSERT_TRUE(txn4 != nullptr);
+    txn4->SetSnapshotOnNextOperation();
+    ASSERT_OK(txn4->SetName("txn4"));
+    // Write to the transaction just to ankor a snapshot.
+    ASSERT_OK(txn4->Put("foofoo4", "bar"));
+    get_perf_context()->Reset();
+    ASSERT_OK(txn4->GetForUpdate(read_options, "foo", &value));
+    // Only active memtable should be checked, and extra read(s)
+    // will be made to fetch the value.
+    if (attempt == kAttemptHistoryMemtable) {
+      ASSERT_EQ(2, get_perf_context()->get_from_memtable_count);
+    } else {
+      assert(attempt == kAttemptImmMemTable);
+      ASSERT_EQ(3, get_perf_context()->get_from_memtable_count);
```

Comment:
Why 3? I was expecting the immutable memtable to be skipped.

Reply:
immutable memtable holds data not yet flushed, and should not be skipped.

Reply:
But txn4 has a snapshot larger than the smallest in the active memtable. Should not the optimization applied in this  patch prevent it from reading the immutable memtable?